cloud-minecraft/cloud-paper/src/main/java/cloud/commandframework/paper/argument/KeyedWorldArgument.java

```diff
@@ -0,0 +1,174 @@
+import java.util.Queue;
+import java.util.function.BiFunction;
+
+/**
+ * cloud argument type that parses Bukkit {@link World worlds} from a namespaced key
+ *
+ * @param <C> Command sender type
+ */
+public class KeyedWorldArgument<C> extends CommandArgument<C, World> {
+
+    protected KeyedWorldArgument(
+            final boolean required,
+            final @NonNull String name,
+            final @NonNull String defaultValue,
+            final @Nullable BiFunction<CommandContext<C>, String, List<String>> suggestionsProvider,
+            final @NonNull ArgumentDescription defaultDescription
+    ) {
+        super(required, name, new KeyedWorldParser<>(), defaultValue, World.class, suggestionsProvider, defaultDescription);
+    }
+
+    /**
+     * Create a new builder
+     *
+     * @param name Name of the argument
+     * @param <C>  Command sender type
+     * @return Created builder
+     */
+    public static <C> KeyedWorldArgument.@NonNull Builder<C> newBuilder(final @NonNull String name) {
+        return new KeyedWorldArgument.Builder<>(name);
+    }
+
+    /**
+     * Create a new required argument
+     *
+     * @param name Argument name
+     * @param <C>  Command sender type
+     * @return Created argument
+     */
+    public static <C> @NonNull KeyedWorldArgument<C> of(final @NonNull String name) {
+        return KeyedWorldArgument.<C>newBuilder(name).asRequired().build();
+    }
+
+    /**
+     * Create a new optional argument
+     *
+     * @param name Argument name
+     * @param <C>  Command sender type
+     * @return Created argument
+     */
+    public static <C> @NonNull KeyedWorldArgument<C> optional(final @NonNull String name) {
+        return KeyedWorldArgument.<C>newBuilder(name).asOptional().build();
+    }
+
+    /**
+     * Create a new optional argument with a default value
+     *
+     * @param name         Argument name
+     * @param defaultValue Default value
+     * @param <C>          Command sender type
+     * @return Created argument
+     */
+    public static <C> @NonNull KeyedWorldArgument<C> optional(
+            final @NonNull String name,
+            final @NonNull String defaultValue
+    ) {
+        return KeyedWorldArgument.<C>newBuilder(name).asOptionalWithDefault(defaultValue).build();
+    }
+
+    public static final class Builder<C> extends CommandArgument.TypedBuilder<C, World, Builder<C>> {
+
+        private Builder(final @NonNull String name) {
+            super(World.class, name);
+        }
+
+        @Override
+        public @NonNull KeyedWorldArgument<C> build() {
+            return new KeyedWorldArgument<>(
+                    this.isRequired(),
+                    this.getName(),
+                    this.getDefaultValue(),
+                    this.getSuggestionsProvider(),
+                    this.getDefaultDescription()
+            );
+        }
+
+    }
+
+    public static final class KeyedWorldParser<C> implements ArgumentParser<C, World> {
```

Comment:
This can just be called `Parser` similar to `Builder`, having the argument name here is kinda redundant as it's an inner class